urllib3/util/rfc6555.py

```diff
@@ -0,0 +1,223 @@
+                cache_entry = (sock.family,
+                               current_time() + _HAPPY_EYEBALLS_CACHE_TIME)
+                _HAPPY_EYEBALLS_CACHE[self._address] = cache_entry
+
+            return sock
+        finally:
+            self._cleanup()
+
+    def _algorithm(self, family=None):
+        """ Executes the algorithm for a given family
+        or no family if IPv4 or IPv6 are allowed. """
+        if family is None:
+            family = socket.AF_UNSPEC
+        host, port = self._address
+        dns_results = socket.getaddrinfo(host, port, family, socket.SOCK_STREAM)
+        dns_results = list(filter(lambda *args: args[0] not in _IP_FAMILIES,
+                                  dns_results))
+
+        result = None
+        for family, socktype, proto, _, sa in dns_results:
+            self._create_socket(family, socktype, proto, sa)
+            result = self._wait_for_connection(False)
+            if result:
+                break
+            self._timeout_check()
+
+        if result is None:
+            if self._error:
+                raise self._error
+            else:
+                raise socket.error(errno.ECONNREFUSED)
+
+        # Restore the old timeout value and remove it from cleanup.
+        result.settimeout(self._timeout)
+        self._sockets.remove(result)
+        self._selector.unregister(result)
+
+        return result
+
+    def _wait_for_connection(self, last_socket=False):
+        """ Does the actual selecting on the sockets waiting
+        for one to be ready to be written to. """
+        # Make sure we don't have any errored sockets.
+        for sock in self._sockets:
+            self._socket_check(sock)
+
+        # If there are no sockets to select on, then don't try.
+        if not self._sockets:
+            return None
+
+        # If it's the last socket then we can block until timeout
+        if last_socket:
+            if self._timeout is None:
+                select_time = None
+            else:
+                select_time = max(0.0, self._timeout - (current_time() - self._start_time))
+        # Otherwise block for 200ms like the RFC suggests.
+        else:
+            select_time = self._get_select_time()
+
+        ready = self._selector.select(select_time)
+
+        for key, _ in ready:
+            sock = key.fileobj
+
+            # Successfully found a proper connected socket.
+            if self._socket_check(sock):
+                return sock
+
+        return None
+
+    def _create_socket(self, family, socktype, proto, sa):
+        """ Tries it's best to create a socket and register
+        the new socket with the selector. If it fails, no
+        problem, the algorithm continues as usual. """
+        sock = None
+        try:
+            sock = socket.socket(family, socktype, proto)
+            if self._timeout is socket._GLOBAL_DEFAULT_TIMEOUT:
+                self._timeout = sock.gettimeout()
```

Comment:
If the timeout is equal to the global default socket timeout value then we need to get that value from an actual socket after it's created so we have a usable value later on in case it's been changed.